### PR TITLE
GH1184: Added support for manually including/excluding files for tools

### PR DESCRIPTION
**Pull request by patriksvensson**
- [x] Query string should support multiple parameters with same name
- [x] Add include/exclude parameter
- [x] Better error messages
- [x] Tests

### CODE REVIEW

File: src/Cake.NuGet/NuGetContentResolver.cs [hunk 1]
```diff
@@ -31,19 +40,56 @@ public IReadOnlyCollection<IFile> GetFiles(DirectoryPath path, PackageType type)
...
+            {
+                collection.Add(_globber.GetFiles(pattern));
+            }
+
+            // Include files.
+            if (package.Parameters.ContainsKey("include"))
+            {
+                foreach (var include in package.Parameters["include"])
+                {
+                    collection.Add(_globber.GetFiles(include));
```
Comment (patriksvensson) [inline src/Cake.NuGet/NuGetContentResolver.cs]: Add `path` to include pattern.

File: src/Cake.NuGet/NuGetContentResolver.cs [hunk 2]
```diff
@@ -31,19 +40,56 @@ public IReadOnlyCollection<IFile> GetFiles(DirectoryPath path, PackageType type)
...
+                }
+            }
+
+            // Exclude files.
+            if (package.Parameters.ContainsKey("exclude"))
+            {
+                foreach (var exclude in package.Parameters["exclude"])
+                {
+                    collection.Remove(_globber.GetFiles(exclude));
```
Comment (patriksvensson) [inline src/Cake.NuGet/NuGetContentResolver.cs]: Add `path` to exclude pattern.

File: src/Cake.Core/Extensions/UriExtensions.cs [hunk 3]
```diff
@@ -4,36 +4,46 @@
...
             if (uri != null)
             {
                 var query = uri.Query.TrimStart('?');
                 var parts = query.Split(new[] { '&' }, StringSplitOptions.RemoveEmptyEntries);
                 foreach (var part in parts)
                 {
                     var pair = part.Split(new[] { '=' }, StringSplitOptions.None);
-                    switch (pair.Length)
+
+                    if (pair.Length == 1)
```
Comment (patriksvensson) [inline src/Cake.Core/Extensions/UriExtensions.cs]: This is ugly... Fix this.

**Comment by patriksvensson** [comment]
@cake-build/team Ok, I *think* this is ready for review.

**Comment by gep13** [comment]
Relates to #1184

**Comment by gep13** [comment]
@patriksvensson are you in a position to rebase this?  And I can try to take a look tonight?

**Comment by patriksvensson** [comment]
@gep13 Sure, I will do this during the day.

**Comment by patriksvensson** [comment]
@gep13 Rebased

File: src/Cake.NuGet.Tests/Unit/NuGetContentResolverTests.cs [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+
+                // Then
+                Assert.Equal(1, files.Count);
+                Assert.Equal("/Working/tools/Foo/Foo.dll", files.First().Path.FullPath);
+            }
+
+            [Theory]
+            [InlineData("nuget:?package=Foo&include=/**/*.XML")]
+            [InlineData("nuget:?package=Foo&include=**/*.XML")]
+            [InlineData("nuget:?package=Foo&include=./**/*.XML")]
```
Comment (gep13) [inline src/Cake.NuGet.Tests/Unit/NuGetContentResolverTests.cs]: @patriksvensson I am assuming that the include/exclude is not case sensitive, right?

Comment (patriksvensson) [inline src/Cake.NuGet.Tests/Unit/NuGetContentResolverTests.cs]: Good question. Will add some tests :smile:

Comment (patriksvensson) [inline src/Cake.NuGet.Tests/Unit/NuGetContentResolverTests.cs]: The include/exclude is case sensitive due to the globber. How should we proceed?

Comment (patriksvensson) [inline src/Cake.NuGet.Tests/Unit/NuGetContentResolverTests.cs]: We could change the interface for `IGlobber` to take a `PathComparer` and move the existing method to an extension method (that uses the default comparer). A breaking change, but something that would be useful in general. Or we could extend globbing to accept regex expressions, like here: https://en.wikipedia.org/wiki/Glob_(programming)

Comment (gep13) [inline src/Cake.NuGet.Tests/Unit/NuGetContentResolverTests.cs]: @patriksvensson hmm, I think Regex might be over complicating it.  I think I would prefer to add a breaking change, where the user can control the outcome, without having to dig into regex.  @devlead thoughts?

Comment (devlead) [inline src/Cake.NuGet.Tests/Unit/NuGetContentResolverTests.cs]: I'll say ship case sensitive. And if necessary break next release.

Comment (gep13) [inline src/Cake.NuGet.Tests/Unit/NuGetContentResolverTests.cs]: @devlead I would be ok with that as well.

Comment (patriksvensson) [inline src/Cake.NuGet.Tests/Unit/NuGetContentResolverTests.cs]: @devlead @gep13 Ok, sounds good to me :smile:

**Comment by patriksvensson** [comment]
@gep13 Rebased against upstream/develop.

**Pull request merged**